content/file/utils.go

```diff
@@ -166,29 +167,41 @@ func extractTarDirectory(dir, prefix string, r io.Reader, buf []byte) error {
 		}
 
 		// Name check
-		name := header.Name
-		path, err := ensureBasePath(dir, prefix, name)
+		filename := header.Name
+		filePathRel, err := resolveRelToBase(dirPath, dirName, filename)
 		if err != nil {
 			return err
 		}
-		path = filepath.Join(dir, path)
+		filePath := filepath.Join(dirPath, filePathRel)
 
 		// Create content
 		switch header.Typeflag {
 		case tar.TypeReg:
-			err = writeFile(path, tr, header.FileInfo().Mode(), buf)
+			err = writeFile(filePath, tr, header.FileInfo().Mode(), buf)
 		case tar.TypeDir:
-			err = os.MkdirAll(path, header.FileInfo().Mode())
+			err = os.MkdirAll(filePath, header.FileInfo().Mode())
 		case tar.TypeLink:
+			// NOTE: ORAS does not generate hard links when creating tarballs.
+			// If a hard link is found in the tarball, it will be extracted.
+			// If the target link already exists, os.Link will throw an error.
+			// This is a known limitation and will not be addressed.
 			var target string
-			if target, err = ensureLinkPath(dir, prefix, path, header.Linkname); err == nil {
-				err = os.Link(target, path)
+			if target, err = ensureLinkPath(dirPath, dirName, filePath, header.Linkname); err == nil {
+				err = os.Link(target, filePath)
 			}
 		case tar.TypeSymlink:
 			var target string
-			if target, err = ensureLinkPath(dir, prefix, path, header.Linkname); err == nil {
-				err = os.Symlink(target, path)
+			target, err = ensureLinkPath(dirPath, dirName, filePath, header.Linkname)
+			if err != nil {
+				return err
+			}
+			if _, err := os.Lstat(filePath); err == nil {
+				// link already exists, remove it first
+				if err := os.Remove(filePath); err != nil {
+					return err
+				}
 			}
+			err = os.Symlink(target, filePath)
```

Comment:
We should assume the link does not exist. If `os.Symlink()` fails, we should then remove and retry.

Reply:
This makes sense, updated.